automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookPage.scala

```diff
@@ -183,24 +183,45 @@ class NotebookPage(override val url: String)(override implicit val authToken: Au
     click on kernelMenu
     click on (await enabled restartKernelSelection)
     click on (await enabled restartKernelConfirmationSelection)
-    await condition (isKernelReady, timeout.toSeconds)
+    waitForIdleKernel(timeout)
   }
 
   /**
-    * wait for kernel to become ready. default timeout == 2.minutes
+    * wait for kernel icon to become idle. default timeout == 1.minute
     */
-  def awaitReadyKernel(timeout: FiniteDuration = 2.minutes): Unit = {
+  def waitForIdleKernel(timeout: FiniteDuration = 1.minutes): Unit = {
     await condition (
-      !cellsAreRunning &&
-      find(id("notification_kernel")).exists(_.underlying.getCssValue("display") == "none") &&
-      find(id("kernel_indicator_icon")).exists(_.underlying.getAttribute("class") == "kernel_idle_icon"), timeout.toSeconds
+      !cellsAreRunning && isKernelIdle() &&
+        find(id("notification_kernel")).exists(_.underlying.getCssValue("display") == "none"), timeout.toSeconds
     )
-    await condition (find(id("kernel_indicator_icon")).exists(_.isDisplayed), 5) // adds little time to check isDisplayed to prevent StaleWebElementException
+    await condition (find(id("kernel_indicator_icon")).exists(_.isDisplayed), 2) // little time to check isDisplayed to prevent StaleWebElementException
+  }
+
+  /**
+    * wait for kernel to become ready. default timeout == 1.minute
+    */
+  def awaitReadyKernel(timeout: FiniteDuration = 1.minute): Unit = {
+    try {
+      waitForIdleKernel(timeout)
+    } catch {
+      case _ =>
+        if (isKernelDisconnected()) {
+          restartKernel()
+        }
+    }
   }
 
   def clickRunCell(timeout: FiniteDuration = 2.minutes): Unit = {
     click on runCellButton
     awaitReadyKernel(timeout)
   }
 
+  def isKernelDisconnected(): Boolean = {
+    find(id("kernel_indicator_icon")).exists(_.underlying.getAttribute("class") == "kernel_disconnected_icon")
+  }
+
+  def isKernelIdle(): Boolean = {
```

Comment:
idle kernel is ready kernel. text `kernel ready` is not reliable way to check for kernel ready. sometimes text is displayed when kernel is still `busy`

Reply:
Would it be more self-explanatory to name the method `isKernelReady()` then..since that's what we care about for our purposes?